pubsub/azure/eventhubs/eventhubs.go

```diff
@@ -15,709 +15,618 @@ package eventhubs
+	// backOff max retry config is set to 3, which means 3 retries by default
+	aeh.backOffConfig = retry.DefaultConfig()
+	aeh.backOffConfig.MaxRetries = 3
+	err = retry.DecodeConfigWithPrefix(&aeh.backOffConfig, metadata.Properties, "backOff")
 	if err != nil {
-		return "", err
+		return fmt.Errorf("failed to decode backoff configuration")
 	}
-	return parsed.HubName, nil
+
+	return nil
 }
 
-func (aeh *AzureEventHubs) ensureEventHub(ctx context.Context, hubName string) error {
-	if aeh.hubManager == nil {
-		aeh.logger.Errorf("hubManager client not initialized properly.")
-		return fmt.Errorf("hubManager client not initialized properly")
-	}
-	entity, err := aeh.getHubEntity(ctx, hubName)
-	if err != nil {
-		return err
-	}
-	if entity == nil {
-		if err := aeh.createHubEntity(ctx, hubName); err != nil {
-			return err
-		}
-	}
+func (aeh *AzureEventHubs) Features() []pubsub.Feature {
 	return nil
 }
 
-func (aeh *AzureEventHubs) ensureSubscription(ctx context.Context, hubName string) error {
-	err := aeh.ensureEventHub(ctx, hubName)
-	if err != nil {
-		return err
-	}
-	_, err = aeh.getConsumerGroupsClient()
-	if err != nil {
-		return err
+// Publish sends a message to Azure Event Hubs.
+func (aeh *AzureEventHubs) Publish(ctx context.Context, req *pubsub.PublishRequest) error {
+	if req.Topic == "" {
+		return errors.New("parameter 'topic' is required")
 	}
-	return aeh.createConsumerGroup(ctx, hubName)
-}
 
-func (aeh *AzureEventHubs) getConsumerGroupsClient() (*mgmt.ConsumerGroupsClient, error) {
-	if aeh.cgClient != nil {
-		return aeh.cgClient, nil
+	// Get the partition key and create the batch of messages
+	batchOpts := &azeventhubs.EventDataBatchOptions{}
+	if pk := req.Metadata["partitionKey"]; pk != "" {
+		batchOpts.PartitionKey = &pk
 	}
-	client := mgmt.NewConsumerGroupsClientWithBaseURI(aeh.managementSettings.AzureEnvironment.ResourceManagerEndpoint,
-		aeh.metadata.SubscriptionID)
-	a, err := aeh.managementSettings.GetAuthorizer()
-	if err != nil {
-		return nil, err
+	messages := []*azeventhubs.EventData{
+		{
+			Body:        req.Data,
+			ContentType: req.ContentType,
+		},
 	}
-	client.Authorizer = a
-	aeh.cgClient = &client
-	return aeh.cgClient, nil
+
+	// Publish the message
+	return aeh.doPublish(ctx, req.Topic, messages, batchOpts)
 }
 
-func (aeh *AzureEventHubs) createConsumerGroup(parentCtx context.Context, hubName string) error {
-	create := false
-	backOffConfig := retry.DefaultConfig()
-	backOffConfig.Policy = retry.PolicyExponential
-	backOffConfig.MaxInterval = resourceCheckMaxRetryInterval
-	backOffConfig.MaxRetries = resourceCheckMaxRetry
+// BulkPublish sends data to Azure Event Hubs in bulk.
+func (aeh *AzureEventHubs) BulkPublish(ctx context.Context, req *pubsub.BulkPublishRequest) (pubsub.BulkPublishResponse, error) {
+	res := pubsub.BulkPublishResponse{}
 
-	b := backOffConfig.NewBackOffWithContext(parentCtx)
+	if req.Topic == "" {
+		return res, errors.New("parameter 'topic' is required")
```

Comment:
do we need to send a response of all failures so that it is retried again by dapr?

Reply:
Yes you're right. Let me fix that